File: android/app/src/main/java/dev/ejunz/node/network/NodeApi.kt
package dev.ejunz.node.network

import android.util.Base64
import dev.ejunz.node.model.NodeCredentials
import dev.ejunz.node.model.NodeDevice
import dev.ejunz.node.model.ZigbeeStatus
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder

class NodeApi(private val credentials: NodeCredentials) {
    private val baseUrl = credentials.baseUrl.trimEnd('/').let {
        if (it.startsWith("http://") || it.startsWith("https://")) it else "http://$it"
    }
    private val authorization = "Basic " + Base64.encodeToString(
        "${credentials.username}:${credentials.password}".toByteArray(),
        Base64.NO_WRAP,
    )

    suspend fun status(): ZigbeeStatus = withContext(Dispatchers.IO) {
        val json = request("/zigbee2mqtt/status")
        ZigbeeStatus(
            connected = json.optBoolean("connected"),
            error = json.optString("error").ifBlank { null },
            devicesCached = json.optInt("devicesCached").takeIf { json.has("devicesCached") },
        )
    }

    suspend fun devices(): List<NodeDevice> = withContext(Dispatchers.IO) {
        val array = request("/zigbee2mqtt/devices").optJSONArray("devices") ?: return@withContext emptyList()
        List(array.length()) { index ->
            val item = array.getJSONObject(index)
            val state = mutableMapOf<String, String>()
            item.optJSONObject("state")?.keys()?.forEach { key -> state[key] = item.optJSONObject("state")?.optString(key).orEmpty() }
            NodeDevice(
                deviceId = item.optString("deviceId"),
                friendlyName = item.optString("friendlyName", item.optString("deviceId")),
                model = item.optString("model").ifBlank { null },
                vendor = item.optString("vendor").ifBlank { null },
                online = if (item.has("online")) item.optBoolean("online") else null,
                supportsOnOff = if (item.has("supportsOnOff")) item.optBoolean("supportsOnOff") else null,
                currentState = item.optString("currentState").ifBlank { null },
                state = state,
            )
        }
    }

    suspend fun setDeviceState(deviceId: String, state: String) = withContext(Dispatchers.IO) {
        request("/zigbee2mqtt/device/${URLEncoder.encode(deviceId, "UTF-8")}", "POST", JSONObject().put("state", state))
    }

    private fun request(path: String, method: String = "GET", body: JSONObject? = null): JSONObject {
        val connection = (URL("$baseUrl$path").openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = 10_000
            readTimeout = 10_000
            setRequestProperty("Accept", "application/json")
            setRequestProperty("Authorization", authorization)
            if (body != null) {
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
            }
        }
        body?.toString()?.toByteArray()?.let { connection.outputStream.use { stream -> stream.write(it) } }
        val response = (if (connection.responseCode in 200..299) connection.inputStream else connection.errorStream)
            ?.bufferedReader()?.use { it.readText() }.orEmpty()
        if (connection.responseCode !in 200..299) error("Node HTTP ${connection.responseCode}: $response")
        return JSONObject(response.ifBlank { "{}" })
    }
}

File: src/api/nodeApi.ts
import type { DevicesResponse, NodeCredentials, ZigbeeStatus } from '../types';

function normalizeBaseUrl(value: string): string {
  const trimmed = value.trim().replace(/\/+$/, '');
  if (!/^https?:\/\//i.test(trimmed)) return `http://${trimmed}`;
  return trimmed;
}

function basicAuth(credentials: NodeCredentials): string {
  return `Basic ${btoa(`${credentials.username}:${credentials.password}`)}`;
}

export class NodeApi {
  readonly baseUrl: string;

  constructor(private readonly credentials: NodeCredentials) {
    this.baseUrl = normalizeBaseUrl(credentials.baseUrl);
  }

  private async request<T>(path: string, init: RequestInit = {}): Promise<T> {
    const response = await fetch(`${this.baseUrl}${path}`, {
      ...init,
      headers: {
        Accept: 'application/json',
        Authorization: basicAuth(this.credentials),
        ...(init.body ? { 'Content-Type': 'application/json' } : {}),
        ...init.headers,
      },
    });

    const text = await response.text();
    let payload: unknown = null;
    try {
      payload = text ? JSON.parse(text) : null;
    } catch {
      payload = text;
    }

    if (!response.ok) {
      const message = typeof payload === 'object' && payload !== null && 'error' in payload
        ? String((payload as { error?: unknown }).error)
        : `Node returned HTTP ${response.status}`;
      throw new Error(message);
    }

    return payload as T;
  }

  async testConnection(): Promise<ZigbeeStatus> {
    return this.request<ZigbeeStatus>('/zigbee2mqtt/status');
  }

  async getStatus(): Promise<ZigbeeStatus> {
    return this.request<ZigbeeStatus>('/zigbee2mqtt/status');
  }

  async listDevices(): Promise<DevicesResponse> {
    return this.request<DevicesResponse>('/zigbee2mqtt/devices');
  }

  async setDeviceState(deviceId: string, state: 'ON' | 'OFF' | 'TOGGLE') {
    return this.request<{ ok?: number; error?: string }>(
      `/zigbee2mqtt/device/${encodeURIComponent(deviceId)}`,
      {
        method: 'POST',
        body: JSON.stringify({ state }),
      },
    );
  }
}

export { normalizeBaseUrl };
